gdal/frmts/fits/fitsdataset.cpp

```diff
@@ -331,38 +422,58 @@ CPLErr FITSDataset::Init(fitsfile* hFITS_, bool isExistingFile_) {
     return CE_Failure;
   }
 
-  // The cftsio driver automatically rescales data on read and write
-  // if BSCALE and BZERO are defined as header keywords. This behavior
-  // causes overflows with GDAL and is slightly mysterious, so we
-  // disable this rescaling here.
-  fits_set_bscale(hFITS, 1.0, 0.0, &status);
-
   // Get the image info for this dataset (note that all bands in a FITS dataset
   // have the same type)
   int bitpix;
   int naxis;
   const int maxdim = 3;
   long naxes[maxdim];
+  double offset;
+  
   fits_get_img_param(hFITS, maxdim, &bitpix, &naxis, naxes, &status);
   if (status) {
     CPLError(CE_Failure, CPLE_AppDefined,
              "Couldn't determine image parameters of FITS file %s (%d).",
              GetDescription(), status);
     return CE_Failure;
   }
+  fits_read_key(hFITS, TDOUBLE, "BZERO", &offset, nullptr, &status);
+  if( status )
+  {
+    status = 0;
+    offset = 0.;
+  }
 
-  // Determine data type
+  fits_read_key(hFITS, TDOUBLE, "BLANK", &dfNoDataValue, nullptr, &status);
+  bNoDataSet = !status;
+  status = 0;
+
+  // Determine data type and nodata value if BLANK keyword is absent
   if (bitpix == BYTE_IMG) {
      gdalDataType = GDT_Byte;
      fitsDataType = TBYTE;
   }
   else if (bitpix == SHORT_IMG) {
-    gdalDataType = GDT_Int16;
-    fitsDataType = TSHORT;
+    if (offset == 32768.)
+    {
+      gdalDataType = GDT_UInt16;
+      fitsDataType = TUSHORT;
+    }
+    else {
+      gdalDataType = GDT_Int16;
+      fitsDataType = TSHORT;
+    }
   }
   else if (bitpix == LONG_IMG) {
-    gdalDataType = GDT_Int32;
-    fitsDataType = TINT;
+    if (offset == 2147483648.)
```

Comment:
Is this representable exactly as a double?  Seems like asking for trouble.

Reply:
FITS always stores 16bit and 32bit integers as signed, in order to support unsigned BZERO keyword is used. For 32bit integers BZERO is set to 2147483648. (2^31) (see e.g. https://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/c_user/node23.html)... after some bibliography... :) ...the limit for integer to be exactly represented by double precision is 2^53 (https://en.wikipedia.org/wiki/Floating-point_arithmetic#IEEE_754:_floating_point_in_modern_computers)... so I think we're good, right?